### ISSUE TITLE
Token is not refreshing once it expires

### ISSUE THREAD

**Issue by shree-gjn** [AzureAD/microsoft-authentication-library-for-js#6943]
### Core Library

MSAL.js (@azure/msal-browser)

### Core Library Version

2.31.0

### Wrapper Library

MSAL React (@azure/msal-react)

### Wrapper Library Version

none

### Public or Confidential Client?

Public

### Description

I have two tokens, 

i) User Token
ii) Management Token

Both tokens are not refreshing once my token is expired. 

### Network Trace (Preferrably Fiddler)

- [ ] Sent
- [X] Pending

### MSAL Configuration

```javascript
export const msalConfig: Configuration = {
	auth: {
		 clientId: "clientId",
		 authority: "https://login.microsoftonline.com/common/",
	},
};
```

### Relevant Code Snippets

```javascript
const refreshTokensIfNeeded = async (account) => {
        // Define the scopes you need for your tokens
        const userTokenRequest = {
            scopes: ["user.read", "email", "mailboxsettings.read", "calendars.readwrite", "files.readwrite", "subscription.read.all"],
            account: account,
            forceRefresh: true, // Set to true if you want to force a refresh
        };
    
        const managementTokenRequest = {
            scopes: ["https://management.azure.com/user_impersonation"],
            account: account,
            forceRefresh: true, // Set to true if you want to force a refresh
        };
    
        try {
            // Attempt to acquire the user token silently
            const userTokenResp = await instance.acquireTokenSilent(userTokenRequest).catch(async error => {
                if (error instanceof InteractionRequiredAuthError) {
                    // Fallback to interactive method if silent acquisition fails
                    return await instance.acquireTokenRedirect(userTokenRequest);
                } else {
                    throw error;
                }
            });
    
            // Attempt to acquire the management token silently
            const mgmtResp = await instance.acquireTokenSilent(managementTokenRequest).catch(async error => {
                if (error instanceof InteractionRequiredAuthError) {
                    return await instance.acquireTokenRedirect(managementTokenRequest);
                } else {
                    throw error;
                }
            });
    
            if (userTokenResp && mgmtResp) {
                sessionStorage.setItem("token", userTokenResp.accessToken);
                sessionStorage.setItem("user_token", mgmtResp.accessToken);
    
                // Fetch user profile with the user token
                const userProfile = await AuthService.getUserProfile(userTokenResp.accessToken);
                if (userProfile && userProfile.mail) {
                    sessionStorage.setItem('userProfile', JSON.stringify(userProfile));
                }
    
                // Obtain authentication key using both tokens
                const authReq = {
                    user_access_token: userTokenResp.accessToken,
                    management_access_token: mgmtResp.accessToken
                } as ITokenRequest;
    
                const authKeyResponse = await AuthService.getAuthKey(authReq);
                sessionStorage.setItem('auth_key', authKeyResponse.auth_key);
    
                // Navigate after successful operations
                window.location.href = redirectUrl || '/homepage';
            }
        } catch (error) {
            console.error("Error during token refresh or profile fetching:", error);
            setLoading(false);
            sessionStorage.clear(); // Consider clearing session storage or specific items instead of everything
            toast.error(t("Session Expired. Please Login Again."));
    
            // Optionally, you might want to initiate a login flow here if silent refresh and redirect both fail
        }
    };
```

### Reproduction Steps

While Sign-in, login using popup. 

### Expected Behavior

Token should refresh silently when expires.

### Identity Provider

Entra ID (formerly Azure AD) / MSA

### Browsers Affected (Select all that apply)

Chrome

### Source

Internal (Microsoft)

**Comment by tnorling** [comment]
We'll need a bit more information to provide guidance. Please share the error you're getting when you attempt to renew the tokens & verbose or trace level MSAL logs

**Comment by VB-007** [comment]
Dear Thomas,

Thank you for your reply. I am sharing source code and the error logs.

I'm attaching our source code with debug logs below. 

>         // Define the function to check token expiration and possibly force a refresh
>         const shouldForceRefresh = (expiresOn: string) => {
>             const now = new Date();
>             const expirationTime = new Date(expiresOn);
>             const timeDifference = expirationTime.getTime() - now.getTime();
>             const threshold = 1  60  1000; // 5 minutes before expiration in milliseconds
>            
>             const isRefreshNeeded = timeDifference <= threshold;
>             addLogEntry(`Checking if token refresh is needed. Refresh needed: ${isRefreshNeeded}`);
>             return isRefreshNeeded;
>         };
>    
>         const refreshTokensIfNeeded = async (account, forceRefresh = false) => {
>             addLogEntry(`Attempting to refresh tokens. Force refresh: ${forceRefresh}`);
>             // Define the scopes you need for your tokens
>             const userTokenRequest = {
>                 scopes: ["user.read", "email", "mailboxsettings.read", "calendars.readwrite", "files.readwrite", "subscription.read.all"],
>                 account: account,
>                 forceRefresh: false, // Set to true if you want to force a refresh
>             };
>        
>             const managementTokenRequest = {
>                 scopes: ["https://management.azure.com/user_impersonation"],
>                 account: account,
>                 forceRefresh: false, // Set to true if you want to force a refresh
>             };
>        
>             try {
>                 // Attempt to acquire the user token silently
>                 const userTokenResp = await instance.acquireTokenSilent(userTokenRequest).catch(async error => {
>                     addLogEntry("Interaction required for user token, attempting popup/redirect.");
>                     if (error instanceof InteractionRequiredAuthError) {
>                         instance.acquireTokenPopup(userTokenRequest).then(userTokenResp => {
>                             setAccessToken(userTokenResp.accessToken);
>                         }).catch(e => {
>                             addLogEntry(`Error during user token acquisition via popup: ${e}`);
>                             console.error(e);
>                         });
>                         // Fallback to interactive method if silent acquisition fails
>                         addLogEntry(`Error acquiring user token silently: ${error}`);
>                         return await instance.acquireTokenRedirect(userTokenRequest);
>                     } else {
>                         throw error;
>                     }
>                 });
>        
>                 // Attempt to acquire the management token silently
>                 const mgmtResp = await instance.acquireTokenSilent(managementTokenRequest).catch(async error => {
>                     if (error instanceof InteractionRequiredAuthError) {
>                         return await instance.acquireTokenRedirect(managementTokenRequest);
>                     } else {
>                         throw error;
>                     }
>                 });
>        
>                 if (userTokenResp && mgmtResp) {
>                     sessionStorage.setItem("token", userTokenResp.accessToken);
>                     sessionStorage.setItem("user_token", mgmtResp.accessToken);
>                     const expiresOnString = userTokenResp.expiresOn ? userTokenResp.expiresOn.toISOString() : '';
>                     sessionStorage.setItem("token_expires_on", expiresOnString);
>                     addLogEntry("Tokens refreshed successfully.");
>        
>                     // Fetch user profile with the user token
>                     const userProfile = await AuthService.getUserProfile(userTokenResp.accessToken);
>                     if (userProfile && userProfile.mail) {
>                         sessionStorage.setItem('userProfile', JSON.stringify(userProfile));
>                     }
>        
>                     // Obtain authentication key using both tokens
>                     const authReq = {
>                         user_access_token: userTokenResp.accessToken,
>                         management_access_token: mgmtResp.accessToken
>                     } as ITokenRequest;
>        
>                     const authKeyResponse = await AuthService.getAuthKey(authReq);
>                     sessionStorage.setItem('auth_key', authKeyResponse.auth_key);
>        
>                     // Navigate after successful operations
>                     window.location.href = redirectUrl || '/homepage';
>                 }
>             } catch (error) {
>                 console.error("Error during token refresh or profile fetching:", error);
>                 addLogEntry(`Error during token refresh or profile fetching: ${error}`);
>                 setLoading(false);
>                 sessionStorage.clear(); // Consider clearing session storage or specific items instead of everything
>                 toast.error(t("Session Expired. Please Login Again."));
>             }
>         };
>    
>         const scheduleTokenRefresh = (account) => {
>             addLogEntry("Scheduling token refresh.");
>             const interval = 5  60  1000; // Check every hour
>            
>             setInterval(async () => {
>                 const expiresOnString = sessionStorage.getItem("token_expires_on");
>                 if (expiresOnString && shouldForceRefresh(expiresOnString)) {
>                     // Force refresh if token is near expiration
>                     addLogEntry("Token is near expiration, forcing refresh.");
>                     await refreshTokensIfNeeded(account, true);
>                 }
>             }, interval);
>         };

The below given code is msalConfig,

export const msalConfig: Configuration = {
    auth: {
         clientId: "clientId",
        // authority: "https://login.microsoftonline.com/common/oauth2/v2.0/authorize/",
          authority: "https://login.microsoftonline.com/common/",
    },
    cache: {
        cacheLocation: "localStorage",
      },
};

// Add here scopes for id token to be used at MS Identity Platform endpoints.
export const loginRequest: PopupRequest = {
    scopes: ["user.read", "email", "mailboxsettings.read", "calendars.readwrite", "files.readwrite", "subscription.read.all", "https://management.azure.com/user_impersonation"],
};


// Add here the endpoints for MS Graph API services you would like to use.
export const graphConfig = {
    graphMeEndpoint: "https://graph.microsoft.com/v1.0/me",
}; 


Below given screenshot is the debug log,

￼

So basically it is scheduling token refresh but it is not refreshing. 


Best regards,
Vikas
